README.md
# Factorio-Bland-Ore-Colors

Lets you independently control ore colors ingame and in the minimap, including making them invisible!. Also lets you turn off the minimap onhover patch highlighting. Very useful if you are playing a mixed ore map or just find the ore patches too bright, i.e. dangOreus

Can apply the below color profiles to all the ingame ores and separately to all the ore's minimap colors. Currently only supports base game ores.
Uranium ore has no glow on all color profiles other than "vanilla".

Example Image with Mixed Ore Tiles
![Example of mixed ores](https://thumbs.gfycat.com/CorruptPlaintiveBeauceron.webp)

When the mod is added or the setting changed the visible map will be re-charted (refreshed) so the new map ore colors are applied.

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 20.1.3
Date: 2021-02-14
  Features:
  Changes:
    - Move data changes to the data-update stage to run after other mods may have changed the ore colors. PR by out_ouk.
  Bugfixes:
    - Support ore map color tables being numericly indexed rather than alphabet letter indexed. PR by out_ouk.
  Info:
---------------------------------------------------------------------------------------------------
Version: 20.1.2
Date: 2020-12-05
  Features:
    - Updated to Factorio 1.1.
  Changes:
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 18.1.1
Date: 2020-01-26
  Features:
    - Updated to Factorio 0.18
    - Added separate options for ingame resource entity color profile and minimap color profile
    - Added minimap highlighting option so when hovering over an ore you can stop the whole patch being highlighted.
  Changes:
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2019-02-13
  Features:
    - Initial Release
  Changes:
  Bugfixes:
  Info:

settings.lua
data:extend(
    {
        {
            name = "bland_ore_colors-entity_color_mode",
            type = "string-setting",
            default_value = "dull",
            allowed_values = {"dull", "bland", "grey", "invisible", "vanilla"},
            setting_type = "startup",
            order = "1001"
        },
        {
            name = "bland_ore_colors-map_color_mode",
            type = "string-setting",
            default_value = "dull",
            allowed_values = {"dull", "bland", "grey", "invisible", "vanilla"},
            setting_type = "startup",
            order = "1002"
        },
        {
            name = "bland_ore_colors-enable_map_highlighting",
            type = "bool-setting",
            default_value = "true",
            setting_type = "startup",
            order = "1003"
        }
    }
)


data-updates.lua
local Constants = require("constants")
local Utils = require("utility/utils")
--local Logging = require("utility/logging")

local function GetColorForSetting(oreColor, settingValue)
    if settingValue == "vanilla" then
        return oreColor
    elseif settingValue == "dull" then
        oreColor[4] = 0.8
    elseif settingValue == "bland" then
        local function CalcBlandColor(oreColor)
            local dif = (0.5 - oreColor) / 4 * 3
            return oreColor + dif
        end
        oreColor = {
            CalcBlandColor(oreColor[1] or oreColor.r),
            CalcBlandColor(oreColor[2] or oreColor.g),
            CalcBlandColor(oreColor[3] or oreColor.b),
            0.8
        }
    elseif settingValue == "grey" then
        oreColor = {0.3, 0.3, 0.3, 0.8}
    elseif settingValue == "invisible" then
        oreColor = {0, 0, 0, 0}
    end
    return oreColor
end

local entityColorModeSetting = settings.startup["bland_ore_colors-entity_color_mode"].value
local mapColorModeSetting = settings.startup["bland_ore_colors-map_color_mode"].value
local mapHighlightingSetting = settings.startup["bland_ore_colors-enable_map_highlighting"].value

local oreList = {"iron-ore", "copper-ore", "coal", "stone", "uranium-ore"}
local oreEntityColors, oreMapColors = {}, {}
for _, name in pairs(oreList) do
    oreEntityColors[name] = Utils.DeepCopy(data.raw["resource"][name].map_color)
    oreEntityColors[name] = GetColorForSetting(oreEntityColors[name], entityColorModeSetting)
end
for _, name in pairs(oreList) do
    oreMapColors[name] = Utils.DeepCopy(data.raw["resource"][name].map_color)
    oreMapColors[name] = GetColorForSetting(oreMapColors[name], mapColorModeSetting)
end

for _, oreName in pairs(oreList) do
    local oreEntityColor = oreEntityColors[oreName]
    local oreMapColor = oreMapColors[oreName]
    local oreProto = data.raw["resource"][oreName]

    if entityColorModeSetting ~= "vanilla" then
        local oreSheet = oreProto.stages.sheet
        oreSheet.filename = Constants.AssetModName .. "/graphics/dull-ore.png"
        oreSheet.tint = oreEntityColor
        oreSheet.hr_version.filename = Constants.AssetModName .. "/graphics/hr-dull-ore.png"
        oreSheet.hr_version.tint = oreEntityColor
        oreProto.stages_effect = nil
    end

    if mapColorModeSetting ~= "vanilla" then
        oreProto.map_color = oreMapColor
        if mapColorModeSetting == "invisible" then
            oreProto.map_grid = false
        else
            oreProto.map_grid = true
        end
    end

    if not mapHighlightingSetting then
        oreProto.resource_patch_search_radius = false
    end
end


data.lua
local Constants = require("constants")
local Utils = require("utility/utils")
local Logging = require("utility/logging")

local function GetColorForSetting(oreColor, settingValue)
    if settingValue == "vanilla" then
        return oreColor
    elseif settingValue == "dull" then
        oreColor[4] = 0.8
    elseif settingValue == "bland" then
        local difR = (0.5 - oreColor[1]) / 4 * 3
        oreColor[1] = oreColor[1] + difR
        local difG = (0.5 - oreColor[2]) / 4 * 3
        oreColor[2] = oreColor[2] + difG
        local difB = (0.5 - oreColor[3]) / 4 * 3
        oreColor[3] = oreColor[3] + difB
        oreColor[4] = 0.8
    elseif settingValue == "grey" then
        oreColor = {0.3, 0.3, 0.3, 0.8}
    elseif settingValue == "invisible" then
        oreColor = {0, 0, 0, 0}
    end
    return oreColor
end

local entityColorModeSetting = settings.startup["bland_ore_colors-entity_color_mode"].value
local mapColorModeSetting = settings.startup["bland_ore_colors-map_color_mode"].value
local mapHighlightingSetting = settings.startup["bland_ore_colors-enable_map_highlighting"].value

local oreList = {"iron-ore", "copper-ore", "coal", "stone", "uranium-ore"}
local oreEntityColors, oreMapColors = {}, {}
for _, name in pairs(oreList) do
    oreEntityColors[name] = Utils.DeepCopy(data.raw["resource"][name].map_color)
    oreEntityColors[name] = GetColorForSetting(oreEntityColors[name], entityColorModeSetting)
end
for _, name in pairs(oreList) do
    oreMapColors[name] = Utils.DeepCopy(data.raw["resource"][name].map_color)
    oreMapColors[name] = GetColorForSetting(oreMapColors[name], mapColorModeSetting)
end

for _, oreName in pairs(oreList) do
    Logging.LogPrint(oreName .. ":")
    local oreEntityColor = oreEntityColors[oreName]
    local oreMapColor = oreMapColors[oreName]
    local oreProto = data.raw["resource"][oreName]

    if entityColorModeSetting ~= "vanilla" then
        Logging.LogPrint(Utils.TableContentsToJSON(oreEntityColor, "oreEntityColor"))
        local oreSheet = oreProto.stages.sheet
        oreSheet.filename = Constants.AssetModName .. "/graphics/dull-ore.png"
        oreSheet.tint = oreEntityColor
        oreSheet.hr_version.filename = Constants.AssetModName .. "/graphics/hr-dull-ore.png"
        oreSheet.hr_version.tint = oreEntityColor
        oreProto.stages_effect = nil
    end

    if mapColorModeSetting ~= "vanilla" then
        oreProto.map_color = oreMapColor
        Logging.LogPrint(Utils.TableContentsToJSON(oreMapColor, "oreMapColor"))
        if mapColorModeSetting == "invisible" then
            oreProto.map_grid = false
        else
            oreProto.map_grid = true
        end
    end

    if not mapHighlightingSetting then
        oreProto.resource_patch_search_radius = false
    end
end
